File: typescript/src/hlc.ts
/**
 * HLC-WID (Hybrid Logical Clock WID) generation, validation, and parsing.
 * Format: YYYYMMDDTHHMMSS[mmm].<lcW>Z-<node>[-<padZ>]
 */

import { type TimeUnit, timeDigits } from "./time";

/** Parsed components of an HLC-WID after a successful parse. */
export interface ParsedHlcWid {
  /** Raw HLC-WID string. */
  raw: string;
  /** UTC timestamp portion of the HLC-WID. */
  timestamp: Date;
  /** Logical clock counter embedded in the HLC-WID. */
  logicalCounter: number;
  /** Node identifier appended to the HLC-WID. */
  node: string;
  /** Optional padding when Z > 0. */
  padding: string | null;
}

/** Minimal Hybrid Logical Clock state snapshot. */
export interface HLCState {
  /** Physical timestamp used by the generator. */
  pt: number;
  /** Logical counter component. */
  lc: number;
}

/** Configuration options for `HLCWidGen`. */
export interface HLCWidGenOptions {
  /** Unique node identifier suffix. */
  node: string;
  /** Width of the logical counter (default 4). */
  W?: number;
  /** Optional padding length (default 0). */
  Z?: number;
  /** Time precision (defaults to `sec`). */
  timeUnit?: TimeUnit;
}

/** Cache for HLC-WID regex instances per width/time unit pair. */
const HLC_BASE_RE_CACHE = new Map<string, RegExp>();
/** Cache for hex validation patterns keyed by padding length. */
const HEX_RE_CACHE = new Map<number, RegExp>();
/** Node identifier pattern reused by HLC generators. */
const NODE_RE = /^[A-Za-z0-9_]+$/;

function hlcBaseRe(W: number, unit: TimeUnit): RegExp {
  const key = `${W}:${unit}`;
  const cached = HLC_BASE_RE_CACHE.get(key);
  if (cached) return cached;
  const re = new RegExp(`^(\\d{8})T(\\d{${timeDigits(unit)}})\\.(\\d{${W}})Z-([A-Za-z0-9_]+)(.*)$`);
  HLC_BASE_RE_CACHE.set(key, re);
  return re;
}

function hexRe(Z: number): RegExp {
  const cached = HEX_RE_CACHE.get(Z);
  if (cached) return cached;
  const re = new RegExp(`^[0-9a-f]{${Z}}$`);
  HEX_RE_CACHE.set(Z, re);
  return re;
}

function randomHexChars(Z: number): string {
  if (!globalThis.crypto?.getRandomValues) {
    throw new Error("Secure random generator unavailable in this runtime");
  }
  const bytes = new Uint8Array(Math.ceil(Z / 2));
  globalThis.crypto.getRandomValues(bytes);
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0"))
    .join("")
    .slice(0, Z);
}

function isValidNode(node: string): boolean {
  return NODE_RE.test(node);
}

function parseTimestamp(dateStr: string, timeStr: string, timeUnit: TimeUnit): Date | null {
  const year = parseInt(dateStr.slice(0, 4), 10);
  const month = parseInt(dateStr.slice(4, 6), 10);
  const day = parseInt(dateStr.slice(6, 8), 10);
  const hour = parseInt(timeStr.slice(0, 2), 10);
  const minute = parseInt(timeStr.slice(2, 4), 10);
  const second = parseInt(timeStr.slice(4, 6), 10);
  const millis = timeUnit === "ms" ? parseInt(timeStr.slice(6, 9), 10) : 0;

  if (month < 1 || month > 12) return null;
  if (day < 1 || day > 31) return null;
  if (hour > 23 || minute > 59 || second > 59) return null;
  if (millis < 0 || millis > 999) return null;

  const timestamp = new Date(Date.UTC(year, month - 1, day, hour, minute, second, millis));
  if (isNaN(timestamp.getTime())) return null;
  if (timestamp.getUTCDate() !== day || timestamp.getUTCMonth() + 1 !== month) return null;
  return timestamp;
}

export function validateHlcWid(
  wid: string,
  W = 4,
  Z = 0,
  timeUnit: TimeUnit = "sec"
): boolean {
  return parseHlcWid(wid, W, Z, timeUnit) !== null;
}

export function parseHlcWid(
  wid: string,
  W = 4,
  Z = 0,
  timeUnit: TimeUnit = "sec"
): ParsedHlcWid | null {
  if (W <= 0 || Z < 0) return null;

  const match = hlcBaseRe(W, timeUnit).exec(wid);
  if (!match) return null;

  const [, dateStr, timeStr, lcStr, node, suffixRaw] = match;
  const suffix = suffixRaw ?? "";

  if (!isValidNode(node)) return null;

  const timestamp = parseTimestamp(dateStr, timeStr, timeUnit);
  if (!timestamp) return null;

  const logicalCounter = parseInt(lcStr, 10);

  let padding: string | null = null;
  if (suffix) {
    if (!suffix.startsWith("-")) return null;
    const seg = suffix.slice(1);
    if (Z === 0) return null;
    if (!hexRe(Z).test(seg)) return null;
    padding = seg;
  }

  return { raw: wid, timestamp, logicalCounter, node, padding };
}

/** Generator for HLC-WIDs that keeps the logical counter monotonic. */
export class HLCWidGen {
  private readonly W: number;
  private readonly Z: number;
  private readonly node: string;
  private readonly timeUnit: TimeUnit;
  private readonly maxLC: number;
  private pt = 0;
  private lc = 0;
  private cachedTick = -1;
  private cachedTs = "";

  constructor(options: HLCWidGenOptions) {
    const { node, W = 4, Z = 0, timeUnit = "sec" } = options;
    if (W <= 0) throw new Error("W must be > 0");
    if (Z < 0) throw new Error("Z must be >= 0");
    if (!isValidNode(node)) {
      throw new Error("node must match [A-Za-z0-9_]+");
    }
    this.W = W;
    this.Z = Z;
    this.node = node;
    this.timeUnit = timeUnit;
    this.maxLC = Math.pow(10, W) - 1;
  }

  private nowTick(): number {
    if (this.timeUnit === "ms") {
      return Date.now();
    }
    return Math.floor(Date.now() / 1000);
  }

  private tsForTick(tick: number): string {
    if (tick !== this.cachedTick) {
      this.cachedTick = tick;
      const sec = this.timeUnit === "ms" ? Math.floor(tick / 1000) : tick;
      const ms = this.timeUnit === "ms" ? tick % 1000 : 0;
      const d = new Date(sec * 1000);
      const year = d.getUTCFullYear();
      const mo = String(d.getUTCMonth() + 1).padStart(2, "0");
      const dy = String(d.getUTCDate()).padStart(2, "0");
      const hr = String(d.getUTCHours()).padStart(2, "0");
      const mi = String(d.getUTCMinutes()).padStart(2, "0");
      const sc = String(d.getUTCSeconds()).padStart(2, "0");
      const milli = String(ms).padStart(3, "0");
      this.cachedTs =
        this.timeUnit === "ms"
          ? `${year}${mo}${dy}T${hr}${mi}${sc}${milli}`
          : `${year}${mo}${dy}T${hr}${mi}${sc}`;
    }
    return this.cachedTs;
  }

  private rollover(): void {
    if (this.lc > this.maxLC) {
      this.pt += 1;
      this.lc = 0;
    }
  }

  observe(remotePT: number, remoteLC: number): void {
    if (remotePT < 0 || remoteLC < 0) {
      throw new Error("remote values must be non-negative");
    }
    const now = this.nowTick();
    const newPT = Math.max(now, this.pt, remotePT);

    if (newPT === this.pt && newPT === remotePT) {
      this.lc = Math.max(this.lc, remoteLC) + 1;
    } else if (newPT === this.pt) {
      this.lc += 1;
    } else if (newPT === remotePT) {
      this.lc = remoteLC + 1;
    } else {
      this.lc = 0;
    }
    this.pt = newPT;
    this.rollover();
  }

  next(): string {
    const now = this.nowTick();
    if (now > this.pt) {
      this.pt = now;
      this.lc = 0;
    } else {
      this.lc += 1;
    }
    this.rollover();

    const ts = this.tsForTick(this.pt);
    const lcStr = String(this.lc).padStart(this.W, "0");
    let wid = `${ts}.${lcStr}Z-${this.node}`;
    if (this.Z > 0) {
      wid += `-${randomHexChars(this.Z)}`;
    }
    return wid;
  }

  nextN(n: number): string[] {
    return Array.from({ length: n }, () => this.next());
  }

  get state(): HLCState {
    return { pt: this.pt, lc: this.lc };
  }

  restoreState(pt: number, lc: number): void {
    if (pt < 0 || lc < 0) throw new Error("invalid state");
    this.pt = pt;
    this.lc = lc;
  }
}

export async function asyncNextHlcWid(options: HLCWidGenOptions): Promise<string> {
  return new HLCWidGen(options).next();
}

export async function* asyncHlcWidStream(
  options: HLCWidGenOptions & { count?: number; intervalMs?: number }
): AsyncGenerator<string> {
  const { count = 0, intervalMs = 0, ...genOpts } = options;
  if (count < 0) throw new Error("count must be >= 0");
  if (intervalMs < 0) throw new Error("intervalMs must be >= 0");
  const gen = new HLCWidGen(genOpts);
  let emitted = 0;
  while (count === 0 || emitted < count) {
    yield gen.next();
    emitted++;
    if (intervalMs > 0) {
      await new Promise((resolve) => setTimeout(resolve, intervalMs));
    }
  }
}

File: rust/src/hlc.rs
//! HLC-WID generation and validation.
//!
//! Format: `YYYYMMDDTHHMMSS[mmm].<lcW>Z-<node>[-<padZ>]`

use chrono::{DateTime, Utc};
use rand::random_range;
use regex::Regex;
use std::collections::HashMap;
use std::sync::{LazyLock, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

use crate::wid::{TimeUnit, WidError, parse_timestamp};

/// Parsed HLC-WID components.
#[derive(Debug, Clone, PartialEq)]
pub struct ParsedHlcWid {
    /// The original HLC-WID string.
    pub raw: String,
    /// The embedded UTC timestamp.
    pub timestamp: DateTime<Utc>,
    /// i64, not u32: the spec allows W up to 18, and an all-nines W=10
    /// counter (9999999999) already exceeds u32::MAX.
    pub logical_counter: i64,
    /// Node tag identifying the minting writer.
    pub node: String,
    /// Random hex pad (`None` when Z=0).
    pub padding: Option<String>,
}

// [0-9], never \d: \d is Unicode-aware in this regex engine and the captured
// fields are byte-sliced in parse_ts — multi-byte digits would panic there.
// See the matching comment in wid.rs.
static HLC_PATTERN_W4_Z0_SEC: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^([0-9]{8})T([0-9]{6})\.([0-9]{4})Z-([A-Za-z0-9_]+)$").unwrap());

fn build_pattern(w: usize, z: usize, time_unit: TimeUnit) -> Regex {
    let lc_part = format!(r"([0-9]{{{w}}})");
    let time_digits = match time_unit {
        TimeUnit::Sec => 6,
        TimeUnit::Ms => 9,
    };
    let pad_part = if z > 0 {
        format!(r"(?:-([0-9a-f]{{{z}}}))?$")
    } else {
        r"$".to_string()
    };
    let pattern =
        format!(r"^([0-9]{{8}})T([0-9]{{{time_digits}}})\.{lc_part}Z-([A-Za-z0-9_]+){pad_part}");
    Regex::new(&pattern).unwrap()
}

// Timestamp parsing is shared with the plain-WID parser (crate::wid::
// parse_timestamp); this module used to carry an identical private copy.

/// Compiled-pattern cache for non-default shapes; see the matching cache in
/// wid.rs for the rationale (Regex clones share the compiled program).
static HLC_PATTERN_CACHE: crate::wid::PatternCache = LazyLock::new(|| Mutex::new(HashMap::new()));

fn cached_pattern(w: usize, z: usize, time_unit: TimeUnit) -> Regex {
    let mut cache = HLC_PATTERN_CACHE
        .lock()
        .unwrap_or_else(|poisoned| poisoned.into_inner());
    cache
        .entry((w, z, time_unit))
        .or_insert_with(|| build_pattern(w, z, time_unit))
        .clone()
}

fn is_valid_node(node: &str) -> bool {
    !node.is_empty() && node.chars().all(|c| c.is_ascii_alphanumeric() || c == '_')
}

/// Validate a HLC-WID string for a specific time unit.
pub fn validate_hlc_wid_with_unit(wid: &str, w: usize, z: usize, time_unit: TimeUnit) -> bool {
    parse_hlc_wid_with_unit(wid, w, z, time_unit).is_ok()
}

/// Validate a HLC-WID string in `sec` mode.
pub fn validate_hlc_wid(wid: &str, w: usize, z: usize) -> bool {
    validate_hlc_wid_with_unit(wid, w, z, TimeUnit::Sec)
}

/// Parse an HLC-WID string into its components for a specific time unit.
pub fn parse_hlc_wid_with_unit(
    wid: &str,
    w: usize,
    z: usize,
    time_unit: TimeUnit,
) -> Result<ParsedHlcWid, WidError> {
    if w == 0 || w > crate::wid::MAX_W {
        return Err(WidError::InvalidW);
    }
    if z > crate::wid::MAX_Z {
        return Err(WidError::InvalidZ);
    }

    let pattern = if w == 4 && z == 0 && time_unit == TimeUnit::Sec {
        HLC_PATTERN_W4_Z0_SEC.clone()
    } else {
        cached_pattern(w, z, time_unit)
    };

    let caps = pattern
        .captures(wid)
        .ok_or_else(|| WidError::InvalidFormat(wid.to_string()))?;

    let date_str = &caps[1];
    let time_str = &caps[2];
    let lc_str = &caps[3];
    let node = caps[4].to_string();
    let padding = if z > 0 {
        caps.get(5).map(|m| m.as_str().to_string())
    } else {
        None
    };

    if !is_valid_node(&node) {
        return Err(WidError::InvalidNode);
    }

    let timestamp =
        parse_timestamp(time_unit, date_str, time_str).ok_or(WidError::InvalidTimestamp)?;
    let logical_counter: i64 = lc_str
        .parse()
        .map_err(|_| WidError::InvalidFormat(wid.to_string()))?;

    Ok(ParsedHlcWid {
        raw: wid.to_string(),
        timestamp,
        logical_counter,
        node,
        padding,
    })
}

/// Parse an HLC-WID string in `sec` mode.
pub fn parse_hlc_wid(wid: &str, w: usize, z: usize) -> Result<ParsedHlcWid, WidError> {
    parse_hlc_wid_with_unit(wid, w, z, TimeUnit::Sec)
}

/// HLC generator state.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct HLCState {
    /// Physical time tick (seconds or milliseconds per the generator's unit).
    pub pt: i64,
    /// Logical counter within the physical tick.
    pub lc: i64,
}

/// HLC-WID generator.
pub struct HLCWidGen {
    w: usize,
    z: usize,
    time_unit: TimeUnit,
    node: String,
    max_lc: i64,
    pt: i64,
    lc: i64,
    cached_tick: i64,
    cached_ts: String,
}

impl HLCWidGen {
    /// Create a new HLC-WID generator in `sec` mode.
    pub fn new(node: String, w: usize, z: usize) -> Result<Self, WidError> {
        Self::new_with_time_unit(node, w, z, TimeUnit::Sec)
    }

    /// Create a new HLC-WID generator with a chosen time unit.
    pub fn new_with_time_unit(
        node: String,
        w: usize,
        z: usize,
        time_unit: TimeUnit,
    ) -> Result<Self, WidError> {
        // W > MAX_W would overflow the i64 logical counter (10^19 > i64::MAX).
        if w == 0 || w > crate::wid::MAX_W {
            return Err(WidError::InvalidW);
        }
        if z > crate::wid::MAX_Z {
            return Err(WidError::InvalidZ);
        }
        if !is_valid_node(&node) {
            return Err(WidError::InvalidNode);
        }

        Ok(Self {
            w,
            z,
            time_unit,
            node,
            max_lc: 10_i64.pow(w as u32) - 1,
            pt: 0,
            lc: 0,
            cached_tick: -1,
            cached_ts: String::new(),
        })
    }

    fn current_tick(time_unit: TimeUnit) -> i64 {
        // A pre-1970 system clock yields Err; treat it as tick 0 rather
        // than panicking (the HLC merge logic takes over from there).
        let dur = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default();
        match time_unit {
            TimeUnit::Sec => dur.as_secs() as i64,
            TimeUnit::Ms => dur.as_millis() as i64,
        }
    }

    fn ts_for_tick(&mut self, tick: i64) -> &str {
        // Saturate instead of unwrap-panicking: a corrupted resume state or
        // hostile remote clock could otherwise push the tick out of range.
        let tick = crate::wid::clamp_tick(tick, self.time_unit);
        if tick != self.cached_tick {
            self.cached_tick = tick;
            self.cached_ts = crate::wid::format_tick(tick, self.time_unit);
        }
        &self.cached_ts
    }

    fn rollover_if_needed(&mut self) {
        if self.lc > self.max_lc {
            self.pt += 1;
            self.lc = 0;
        }
    }

    /// Merge remote HLC state.
    pub fn observe(&mut self, remote_pt: i64, remote_lc: i64) -> Result<(), WidError> {
        if remote_pt < 0 || remote_lc < 0 {
            return Err(WidError::InvalidRemoteClock);
        }

        let now = Self::current_tick(self.time_unit);
        let new_pt = now.max(self.pt).max(remote_pt);

        if new_pt == self.pt && new_pt == remote_pt {
            self.lc = self.lc.max(remote_lc) + 1;
        } else if new_pt == self.pt {
            self.lc += 1;
        } else if new_pt == remote_pt {
            self.lc = remote_lc + 1;
        } else {
            self.lc = 0;
        }

        self.pt = new_pt;
        self.rollover_if_needed();
        Ok(())
    }

    /// Generate the next HLC-WID.
    pub fn next_hlc_wid(&mut self) -> String {
        let now = Self::current_tick(self.time_unit);
        if now > self.pt {
            self.pt = now;
            self.lc = 0;
        } else {
            self.lc += 1;
        }
        self.rollover_if_needed();

        let ts = self.ts_for_tick(self.pt).to_string();
        let lc_str = format!("{:0width$}", self.lc, width = self.w);
        let mut wid = format!("{}.{}Z-{}", ts, lc_str, self.node);

        if self.z > 0 {
            const HEX: &[u8; 16] = b"0123456789abcdef";
            let pad: String = (0..self.z)
                .map(|_| HEX[random_range(0..16)] as char)
                .collect();
            wid.push('-');
            wid.push_str(&pad);
        }

        wid
    }

    /// Generate n HLC-WIDs.
    pub fn next_n(&mut self, n: usize) -> Vec<String> {
        self.take(n).collect()
    }

    /// Get current state.
    pub fn state(&self) -> HLCState {
        HLCState {
            pt: self.pt,
            lc: self.lc,
        }
    }

    /// Restore state.
    pub fn restore_state(&mut self, pt: i64, lc: i64) -> Result<(), WidError> {
        if pt < 0 || lc < 0 {
            return Err(WidError::InvalidRemoteClock);
        }
        self.pt = pt;
        self.lc = lc;
        Ok(())
    }

    /// Active time unit.
    pub fn time_unit(&self) -> TimeUnit {
        self.time_unit
    }
}

impl Iterator for HLCWidGen {
    type Item = String;

    fn next(&mut self) -> Option<Self::Item> {
        Some(self.next_hlc_wid())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_validate_hlc_valid() {
        assert!(validate_hlc_wid("20260212T091530.0000Z-node01", 4, 0));
        assert!(validate_hlc_wid(
            "20260212T091530.0042Z-node01-a3f91c",
            4,
            6
        ));
        assert!(validate_hlc_wid_with_unit(
            "20260212T091530123.0042Z-node01-a3f91c",
            4,
            6,
            TimeUnit::Ms
        ));
    }

    #[test]
    fn test_validate_hlc_invalid() {
        assert!(!validate_hlc_wid("20260212T091530.0000Z", 4, 0));
        assert!(!validate_hlc_wid("20260212T091530.0000Z-node-01", 4, 0));
        // Year 0000 is outside the spec range 0001-9999.
        assert!(!validate_hlc_wid("00000101T000000.0000Z-node01", 4, 0));
        assert!(!validate_hlc_wid(
            "20260212T091530.0000Z-node01-ABCDEF",
            4,
            6
        ));
    }

    #[test]
    fn test_unicode_digits_rejected_without_panic() {
        // Same regression as wid.rs: Unicode digits must be a clean reject,
        // not a byte-boundary panic in parse_ts.
        assert!(!validate_hlc_wid("२०२६०२१२T०९१५३०.००००Z-node01", 4, 0));
        assert!(matches!(
            parse_hlc_wid("٢٠٢٦٠٢١٢T٠٩١٥٣٠.٠٠٠٠Z-node01", 4, 0),
            Err(WidError::InvalidFormat(_))
        ));
    }

    #[test]
    fn test_parse_hlc() {
        let p = parse_hlc_wid("20260212T091530.0042Z-node01-a3f91c", 4, 6).unwrap();
        assert_eq!(p.node, "node01");
        assert_eq!(p.logical_counter, 42);
        assert_eq!(p.padding.as_deref(), Some("a3f91c"));

        let p2 =
            parse_hlc_wid_with_unit("20260212T091530123.0042Z-node01-a3f91c", 4, 6, TimeUnit::Ms)
                .unwrap();
        assert_eq!(p2.timestamp.timestamp_subsec_millis(), 123);
    }

    #[test]
    fn test_hlc_monotonic() {
        let mut g = HLCWidGen::new("node01".to_string(), 4, 0).unwrap();
        let a = g.next_hlc_wid();
        let b = g.next_hlc_wid();
        assert!(a < b);
    }

    #[test]
    fn test_hlc_observe() {
        let mut g = HLCWidGen::new("node01".to_string(), 4, 0).unwrap();
        g.observe(10, 5).unwrap();
        let s = g.state();
        assert!(s.pt >= 10);
    }

    #[test]
    fn test_new_rejects_invalid_node_and_w() {
        assert!(matches!(
            HLCWidGen::new("bad-node".to_string(), 4, 0),
            Err(WidError::InvalidNode)
        ));
        assert!(matches!(
            HLCWidGen::new("node01".to_string(), 0, 0),
            Err(WidError::InvalidW)
        ));
    }

    #[test]
    fn test_observe_invalid_remote_clock() {
        let mut g = HLCWidGen::new("node01".to_string(), 4, 0).unwrap();
        assert!(matches!(
            g.observe(-1, 0),
            Err(WidError::InvalidRemoteClock)
        ));
        assert!(matches!(
            g.observe(0, -1),
            Err(WidError::InvalidRemoteClock)
        ));
    }

    #[test]
    fn test_restore_state_invalid() {
        let mut g = HLCWidGen::new("node01".to_string(), 4, 0).unwrap();
        assert!(matches!(
            g.restore_state(-1, 0),
            Err(WidError::InvalidRemoteClock)
        ));
        assert!(matches!(
            g.restore_state(0, -1),
            Err(WidError::InvalidRemoteClock)
        ));
    }

    #[test]
    fn test_next_with_padding_and_next_n() {
        let mut g = HLCWidGen::new("node01".to_string(), 4, 6).unwrap();
        let one = g.next_hlc_wid();
        assert!(one.contains("-node01-"));
        let many = g.next_n(3);
        assert_eq!(many.len(), 3);
        assert!(many[0] < many[1]);
    }

    #[test]
    fn test_parse_hlc_invalid_timestamp() {
        assert!(matches!(
            parse_hlc_wid("20261312T091530.0000Z-node01", 4, 0),
            Err(WidError::InvalidTimestamp)
        ));
    }

    #[test]
    fn test_non_default_w_z_pattern_paths() {
        assert!(validate_hlc_wid("20260212T091530.00042Z-node01-ab", 5, 2));
        let p = parse_hlc_wid("20260212T091530.00042Z-node01-ab", 5, 2).unwrap();
        assert_eq!(p.logical_counter, 42);
        assert_eq!(p.padding.as_deref(), Some("ab"));
    }

    #[test]
    fn test_ms_generator_shape() {
        let mut g =
            HLCWidGen::new_with_time_unit("node01".to_string(), 4, 0, TimeUnit::Ms).unwrap();
        let id = g.next_hlc_wid();
        assert!(validate_hlc_wid_with_unit(&id, 4, 0, TimeUnit::Ms));
    }
}
